README.md
NewPingOSLRF
============

Open Source Laser RangeFinder -OSLRF01 version of Tim Eckels NewPing Library


history.txt
// BACKGROUND:
// When I first received an ultrasonic sensor I was not happy with how poorly
// it worked. Quickly I realized the problem wasn't the sensor, it was the
// available ping and ultrasonic libraries causing the problem. The NewPing
// library totally fixes these problems, adds many new features, and breaths
// new life into these very affordable distance sensors. 
//
// FEATURES:
// * Works with many different ultrasonic sensor models: SR04, SRF05, SRF06, DYP-ME007 & Parallax PING))).
// * Interface with all but the SRF06 sensor using only one Arduino pin.
// * Doesn't lag for a full second if no ping/echo is received.
// * Ping sensors consistently and reliably at up to 30 times per second.
// * Timer interrupt method for event-driven sketches.
// * Built-in digital filter method ping_median() for easy error correction.
// * Uses port registers for a faster pin interface and smaller code size.
// * Allows you to set a maximum distance where pings beyond that distance are read as no ping "clear".
// * Ease of using multiple sensors (example sketch with 15 sensors).
// * More accurate distance calculation (cm, inches & uS).
// * Doesn't use pulseIn, which is slow and gives incorrect results with some ultrasonic sensor models.
// * Actively developed with features being added and bugs/issues addressed.

// HISTORY:
// 08/15/2012 v1.5 - Added ping_median() method which does a user specified
//   number of pings (default=5) and returns the median ping in microseconds
//   (out of range pings ignored). This is a very effective digital filter.
//   Optimized for smaller compiled size (even smaller than skteches that
//   don't use a library).
//
// 07/14/2012 v1.4 - Added support for the Parallax PING))) sensor. Interface
//   with all but the SRF06 sensor using only one Arduino pin. You can also
//   interface with the SRF06 using one pin if you install a 0.1uf capacitor
//   on the trigger and echo pins of the sensor then tie the trigger pin to
//   the Arduino pin (doesn't work with Teensy). To use the same Arduino pin
//   for trigger and echo, specify the same pin for both values. Various bug
//   fixes.
//
// 06/08/2012 v1.3 - Big feature addition, event-driven ping! Uses Timer2
//   interrupt, so be mindful of PWM or timing conflicts messing with Timer2
//   may cause (namely PWM on pins 3 & 11 on Arduino, PWM on pins 9 and 10 on
//   Mega, and Tone library). Simple to use timer interrupt functions you can
//   use in your sketches totaly unrelated to ultrasonic sensors (don't use if
//   you're also using NewPing's ping_timer because both use Timer2 interrupts).
//   Loop counting ping method deleted in favor of timing ping method after
//   inconsistant results kept surfacing with the loop timing ping method.
//   Conversion to cm and inches now rounds to the nearest cm or inch. Code
//   optimized to save program space and fixed a couple minor bugs here and
//   there. Many new comments added as well as line spacing to group code
//   sections for better source readability.
//
// 05/25/2012 v1.2 - Lots of code clean-up thanks to Adruino Forum members.
//   Rebuilt the ping timing code from scratch, ditched the pulseIn code as it
//   doesn't give correct results (at least with ping sensors). The NewPing
//   library is now VERY accurate and the code was simplified as a bonus.
//   Smaller and faster code as well. Fixed some issues with very close ping
//   results when converting to inches. All functions now return 0 only when
//   there's no ping echo (out of range) and a positive value for a successful
//   ping. This can effectively be used to detect if something is out of range
//   or in-range and at what distance. Now compatible with Arduino 0023.
//
// 05/16/2012 v1.1 - Changed all I/O functions to use low-level port registers
//   for ultra-fast and lean code (saves from 174 to 394 bytes). Tested on both
//   the Arduino Uno and Teensy 2.0 but should work on all Arduino-based
//   platforms because it calls standard functions to retrieve port registers
//   and bit masks. Also made a couple minor fixes to defines.
//
// 05/15/2012 v1.0 - Initial release.
// ---------------------------------------------------------------------------



NewPingOSLRF.h
// ---------------------------------------------------------------------------
// PingOSLRF Library - v1.0 - 04/21/2014
//
// Original AUTHOR/LICENSE:
// Created by Tim Eckel - teckel@leethost.com
// Copyright 2012 License: GNU GPL v3 http://www.gnu.org/licenses/gpl-3.0.html
//
// Modified March 2014, for the Open Source Laser Range Finder OSLRF01 
// http://lightware.co.za/shop/en/index.php?controller=attachment&id_attachment=7
// By Michael Ball  unix_guru@hotmail.com 


// LINKS:
// Project home: http://code.google.com/p/arduino-new-ping/
// Blog: http://arduino.cc/forum/index.php/topic,106043.0.html
//
// DISCLAIMER:
// This software is furnished "as is", without technical support, and with no 
// warranty, express or implied, as to its usefulness for any purpose.
//
//
// CONSTRUCTOR:
//   PingOSLRF sonar(trigger_pin, echo_pin [, max_cm_distance])
//     trigger_pin & echo_pin - Arduino pins connected to sensor trigger and echo.
//       NOTE: To use the same Arduino pin for trigger and echo, specify the same pin for both values.
//     max_cm_distance - [Optional] Maximum distance you wish to sense. Default=500cm.
//
// SYNTAX:
//   sonar.ping() - Send a ping and get the echo time (in microseconds) as a result. 
//   sonar.ping_in() - Send a ping and get the distance in whole inches.
//   sonar.ping_cm() - Send a ping and get the distance in whole centimeters.
//   sonar.ping_median(iterations) - Do multiple pings (default=5), discard out of range pings and return median in microseconds. 
//   sonar.convert_in(echoTime) - Convert echoTime from microseconds to inches (rounds to nearest inch).
//   sonar.convert_cm(echoTime) - Convert echoTime from microseconds to centimeters (rounds to nearest cm).
//   sonar.ping_timer(function) - Send a ping and call function to test if ping is complete.
//   sonar.check_timer() - Check if ping has returned within the set distance limit.
//   NewPing::timer_us(frequency, function) - Call function every frequency microseconds.
//   NewPing::timer_ms(frequency, function) - Call function every frequency milliseconds.
//   NewPing::timer_stop() - Stop the timer.
//

#ifndef NewPing_h
#define NewPing_h

#if defined(ARDUINO) && ARDUINO >= 100
	#include <Arduino.h>
#else
	#include <WProgram.h>
	#include <pins_arduino.h>
#endif

#include <avr/io.h>
#include <avr/interrupt.h>

// Shoudln't need to changed these values unless you have a specific need to do so.
#define MAX_SENSOR_DISTANCE 1000 // Maximum sensor distance can be as high as 500cm, no reason to wait for ping longer than sound takes to travel this distance and back.

//********************************** These two need to be worked on.  with a sliding timebase these do not work  *********//
#define US_ROUNDTRIP_IN 722     // Microseconds (uS) it takes sound to travel round-trip 1 inch (2 inches total), uses integer to save compiled code space.
#define US_ROUNDTRIP_CM 1833      // Microseconds (uS) it takes sound to travel round-trip 1cm (2cm total), uses integer to save compiled code space.


// Probably shoudln't change these values unless you really know what you're doing.
#define NO_ECHO 0               // Value returned if there's no ping echo within the specified MAX_SENSOR_DISTANCE or max_cm_distance.
#define MAX_SENSOR_DELAY 25000  // Maximum uS it takes for sensor to start the ping (SRF06 is the highest measured, just under 18ms).
#define ECHO_TIMER_FREQ 24      // Frequency to check for a ping echo (every 24uS is about 0.4cm accuracy).
#define PING_MEDIAN_DELAY 29    // Millisecond delay between pings in the ping_median method.

#define ZERO_THRESHOLD  50	// Analog value to accept as Zero rise.
#define RETURN_THRESHOLD  80	// Analog value to accept as Return rise.




#define FASTADC 1				// Allows ADC Prescalar of 16 or 500khz on 16Mhz Arduino


// defines for setting and clearing register bits
#ifndef bit_get
#define bit_get(p,m) ((p) & (m))
#endif
#ifndef bit_set
#define bit_set(p,m) ((p) |= (m))
#endif
#ifndef bit_clear
#define bit_clear(p,m) ((p) &= ~(m))
#endif
#ifndef bit_flip
#define bit_flip(p,m) ((p) ^= (m))
#endif
#ifndef bit_write
#define bit_write(c,p,m) (c ? bit_set(p,m) : bit_clear(p,m))
#endif
#ifndef BIT
#define BIT(x) (0x01 << (x))
#endif
#ifndef LONGBIT
#define LONGBIT(x) ((unsigned long)0x00000001 << (x)) 
#endif


class NewPing {
	public:
		NewPing(uint8_t sync_pin, uint8_t zero_pin,  uint8_t echo_pin, int max_cm_distance = MAX_SENSOR_DISTANCE);
		unsigned long ping();
		void begin();
		unsigned int ping_in();
		unsigned int ping_cm();
		unsigned long sync_width();
		unsigned long syncWidth;
		unsigned int zero_rise();
		unsigned int zeroRise;
		unsigned int zeroThresh;
		unsigned int echoThresh;
		unsigned int echoValue;
		unsigned long echoWidth;
		unsigned int ping_median(uint8_t it = 5);

		void ping_timer(void (*userFunc)(void));
		boolean check_timer();
		unsigned long ping_result;
		static void timer_us(unsigned int frequency, void (*userFunc)(void));
		static void timer_ms(unsigned long frequency, void (*userFunc)(void));
		static void timer_stop();
	private:
		boolean ping_sync();
		boolean ping_wait_timer();
		uint8_t _syncBit;
		uint8_t _zeroBit;
		uint8_t _echoBit;
		uint8_t _zeroPin;
		uint8_t _echoPin;
		uint8_t _syncPin;
		volatile uint8_t *_syncInput;
		volatile uint8_t *_echoInput;
		volatile uint8_t *_zeroInput;
		unsigned int _maxEchoTime;
		unsigned long _max_time;
		volatile long _syncRise;
		volatile long _syncFall;
		static void timer_setup();
		static void timer_ms_cntdwn();
};


#endif


NewPingOSLRF.cpp
// ---------------------------------------------------------------------------
// Created by Tim Eckel - teckel@leethost.com
// Copyright 2012 License: GNU GPL v3 http://www.gnu.org/licenses/gpl-3.0.html
//
// See "NewPing.h" for purpose, syntax, version history, links, and more.
//
// Modified March 2014, for the Open Source Laser Range Finder OSLRF01 
// http://lightware.co.za/shop/en/index.php?controller=attachment&id_attachment=7
// By Michael Ball  unix_guru@hotmail.com 
// ---------------------------------------------------------------------------

#include "NewPingOSLRF.h"


// ---------------------------------------------------------------------------
// NewPing constructor
// ---------------------------------------------------------------------------

NewPing::NewPing(uint8_t sync_pin, uint8_t zero_pin, uint8_t echo_pin, int max_cm_distance) {
	_syncBit = digitalPinToBitMask(sync_pin);       // Get the port register bitmask for the sync pin.
	_zeroBit = digitalPinToBitMask(zero_pin);       // Get the port register bitmask for the echo pin.
	_syncPin = sync_pin;
	_zeroPin = zero_pin;
	_echoBit = digitalPinToBitMask(echo_pin);       // Get the port register bitmask for the echo pin.
	_echoPin = echo_pin;
	_syncInput = portInputRegister(digitalPinToPort(sync_pin)); 	    // Get the input port register for the sync pin.
	_zeroInput = portInputRegister(digitalPinToPort(zero_pin));         // Get the input port register for the zero pin.
	_echoInput = portInputRegister(digitalPinToPort(echo_pin));         // Get the input port register for the echo pin.

//	_maxEchoTime = min(max_cm_distance, MAX_SENSOR_DISTANCE) * US_ROUNDTRIP_CM + (US_ROUNDTRIP_CM / 2); // Calculate the maximum distance in uS.
	_maxEchoTime = min(max_cm_distance, MAX_SENSOR_DISTANCE) * US_ROUNDTRIP_CM ; // Calculate the 	zeroThresh =  ZERO_THRESHOLD;
	echoThresh =  RETURN_THRESHOLD;

#if FASTADC
  // set ADC prescale to 16
//  bit_set(ADCSRA,ADPS2) ;
//  bit_clear(ADCSRA,ADPS1) ;
//  bit_clear(ADCSRA,ADPS0) ;

  // set ADC prescale to 8
  bit_clear(ADCSRA,ADPS2) ;
  bit_set(ADCSRA,ADPS1) ;
  bit_set(ADCSRA,ADPS0) ;
#endif

}


// ---------------------------------------------------------------------------
// Standard ping methods
// ---------------------------------------------------------------------------

unsigned long NewPing::ping() {
  unsigned long echo_rise;
  unsigned long echo_fall;
  int ev;
	echoValue = 0;					 // Reset Max value of Echo for new reading.
	if (!ping_sync()) return NO_ECHO;                     // Sync a ping, if it returns false, return NO_ECHO.
	while ((echoValue = analogRead(_echoPin)) < echoThresh)
	    if (micros() > _max_time) return NO_ECHO;       // Stop the loop and return NO_ECHO (false) 
							    // if we're beyond the set maximum distance.
	echo_rise = micros();

	while (ev <= echoValue ) {			   // Get maximum height of pulse...
		ev = echoValue;
		echoValue = analogRead(_echoPin);	    
	}
	//echo_fall = micros();
	echoWidth = micros() - echo_rise;   

 	ping_result = (echo_rise - _syncFall ) ; // Calculate ping time, 5uS of overhead.
	return (ping_result);		// ping_result is round trip... just send one way value
}


unsigned int NewPing::ping_in() {
	unsigned long echoTime = NewPing::ping();      // Calls the ping method and returns with the ping echo distance in uS.
	return ((echoTime-zeroRise)*722)/syncWidth; // Convert uS to centimeters.
}


unsigned int NewPing::ping_cm() {
	unsigned long echoTime = NewPing::ping();      // Calls the ping method and returns with the ping echo distance in uS.
	return ((echoTime-zeroRise)*1833)/syncWidth; // Convert uS to centimeters.
}


unsigned int NewPing::ping_median(uint8_t it) {
	unsigned int uS[it], last;
	uint8_t j, i = 0;
	uS[0] = NO_ECHO;
	while (i < it) {
		last = ping();           // Send ping.
		if (last == NO_ECHO) {   // Ping out of range.
			it--;                // Skip, don't include as part of median.
			last = _maxEchoTime; // Adjust "last" variable so delay is correct length.
		} else {                       // Ping in range, include as part of median.
			if (i > 0) {               // Don't start sort till second ping.
				for (j = i; j > 0 && uS[j - 1] < last; j--) // Insertion sort loop.
					uS[j] = uS[j - 1]; // Shift ping array to correct position for sort insertion.
			} else j = 0;              // First ping is starting point for sort.
			uS[j] = last;              // Add last ping to array in sorted position.
			i++;                       // Move to next ping.
		}
		if (i < it) delay(PING_MEDIAN_DELAY - (last >> 10)); // Millisecond delay between pings.
	}
	return (uS[it >> 1]); // Return the ping distance median.
}



// ---------------------------------------------------------------------------
// OSLRF support functions for Constructor.... 
// zero_rise() and sync_width are required to determine start of pulse and 
// time base.   zero_rise initiates outbound laser pulse.  then the difference 
// between that and the start of the return signal is divided by the sync_width
// multiplied by the constant 18.33m        D = ((Rt-Zt) / Sp) * 18.33M
// ---------------------------------------------------------------------------

unsigned long NewPing::sync_width() {		// Calculate the width of the OSLRF01 Sync Pulse
  unsigned long sync_drop;
  unsigned long sync_rise;
	while (!(*_syncInput & _syncBit)) {}                     // Wait for Sync pin to rise so we can find a start state.
	while (*_syncInput & _syncBit) {} 			 // Wait for Sync pin to clear.
	sync_drop = micros();					 // Falling edge time.
	while (!(*_syncInput & _syncBit)) {}                       // Wait for Sync pin to rise.
	sync_rise = micros();					 // Rising edge time.

	syncWidth = (sync_rise - sync_drop) *2 ;
	_maxEchoTime = syncWidth *2; 
	return syncWidth;
}



unsigned int NewPing::zero_rise() {		    // Find the start delay of the outgoing pulse for calculations
	if (!ping_sync()) return NO_ECHO;           // Sync a ping with falling edge of Sync, if it returns false, 
	while (analogRead(_zeroPin) < zeroThresh) {}
	return (micros() - _syncFall);
}

void NewPing::begin(){
	syncWidth = NewPing::sync_width();
	zeroRise = NewPing::zero_rise();
}

// ---------------------------------------------------------------------------
// Standard ping method support functions (not called directly)
// ---------------------------------------------------------------------------

boolean NewPing::ping_sync() {		// Find falling edge of ping sync

	_max_time =  micros() + MAX_SENSOR_DELAY;                  // Set a timeout for the ping to trigger.
	while (!(*_syncInput & _syncBit)) {}                      // Wait for Sync pin to rise so we can find a start state.
	while (*_syncInput & _syncBit && micros() <= _max_time) {} // Wait for Sync pin to clear.
	if (micros() > _max_time) return false;                   // Something went wrong, abort.
	_max_time = micros() + _maxEchoTime;                      // Ping started, set the timeout.
	_syncFall = micros();
	return true;                                              // Ping started successfully.
} 




